### ISSUE TITLE
Attempt to start event loop from atexit() callback.

### ISSUE THREAD

**Issue by GrahamDumpleton** [jupyterhub/jupyterhub#2010]
**Describe the bug**

When sending a ``SIGTERM`` to JupyterHub process, it look like an ``atexit`` register callback is trying to restart the event loop.

**To Reproduce**

Run JupyterHub 0.9.0.

Send JupyterHub process a ``SIGTERM``.

When did this had the following logged.

```

[C 2018-06-25 02:42:09.962 JupyterHub app:1912] Received SIGTERM, shutting down
--
  | Traceback (most recent call last):
  | File "/opt/app-root/bin/jupyterhub", line 4, in <module>
  | main()
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/app.py", line 1951, in launch_instance
  | loop.start()
  | File "/opt/app-root/lib/python3.6/site-packages/tornado/platform/asyncio.py", line 127, in start
  | self.asyncio_loop.run_forever()
  | File "/opt/rh/rh-python36/root/usr/lib64/python3.6/asyncio/base_events.py", line 421, in run_forever
  | self._run_once()
  | File "/opt/rh/rh-python36/root/usr/lib64/python3.6/asyncio/base_events.py", line 1390, in _run_once
  | event_list = self._selector.select(timeout)
  | File "/opt/rh/rh-python36/root/usr/lib64/python3.6/selectors.py", line 445, in select
  | fd_event_list = self._epoll.poll(timeout, max_ev)
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/app.py", line 1914, in sigterm
  | self.atexit()
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/app.py", line 1927, in atexit
  | loop.run_sync(self.cleanup)
  | File "/opt/app-root/lib/python3.6/site-packages/tornado/ioloop.py", line 576, in run_sync
  | self.start()
  | File "/opt/app-root/lib/python3.6/site-packages/tornado/platform/asyncio.py", line 127, in start
  | self.asyncio_loop.run_forever()
  | File "/opt/rh/rh-python36/root/usr/lib64/python3.6/asyncio/base_events.py", line 411, in run_forever
  | 'Cannot run the event loop while another loop is running')
  | RuntimeError: Cannot run the event loop while another loop is running
```

**Expected behavior**

Should shutdown cleanly with no errors.

**Comment by GrahamDumpleton** [comment]
Tornado version is:

```
tornado==5.0.2
```

**Comment by GrahamDumpleton** [comment]
Pinning to:

```
tornado==4.5.3
```

as mentioned in similar issue related to Jupyter notebooks at https://github.com/jupyter/notebook/issues/3397 avoids problem.

**Comment by GrahamDumpleton** [comment]
Alas, you possibly can't use older Tornado version as it appears to break other things.

```
[I 2018-06-25 04:30:56.699 JupyterHub reflector:129] watching for pods with label selector heritage=jupyterhub,component=singleuser-server in namespace myproject
  | [E 2018-06-25 04:30:56.719 JupyterHub user:474] Unhandled error starting grumpy's server: Task got bad yield: <tornado.concurrent.Future object at 0x7fd001742b70>
  | [E 2018-06-25 04:30:56.726 JupyterHub user:483] Failed to cleanup grumpy's server that failed to start
  | Traceback (most recent call last):
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/user.py", line 479, in spawn
  | await self.stop()
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/user.py", line 546, in stop
  | status = await spawner.poll()
  | File "<string>", line 3, in __await__
  | RuntimeError: Task got bad yield: <tornado.concurrent.Future object at 0x7fd001742c88>
```

That or managed to get database into a strange state by reverting back to older Tornado 4 while using same database. When rolled forward again to Tornado 5, didn't get this error and cleanup occurred okay.

**Comment by GrahamDumpleton** [comment]
Additional related exception traceback.

```
Traceback (most recent call last):
  | File "/opt/app-root/lib/python3.6/site-packages/tornado/web.py", line 1512, in _execute
  | result = yield result
  | File "<string>", line 6, in _wrap_awaitable
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 999, in get
  | await self.spawn_single_user(user)
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 670, in spawn_single_user
  | await gen.with_timeout(timedelta(seconds=self.slow_spawn_timeout), finish_spawn_future)
  | File "<string>", line 3, in __await__
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/handlers/base.py", line 625, in finish_user_spawn
  | await spawn_future
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/user.py", line 486, in spawn
  | raise e
  | File "/opt/app-root/lib/python3.6/site-packages/jupyterhub/user.py", line 406, in spawn
  | url = await gen.with_timeout(timedelta(seconds=spawner.start_timeout), f)
  | File "<string>", line 3, in __await__
  | RuntimeError: Task got bad yield: <tornado.concurrent.Future object at 0x7f2ff8fa8fd0>
```

**Comment by minrk** [comment]
tornado 5 is explicitly required as a dependency of jupyterhub 0.9, so I'd ignore any of the tornado 4 errors and focus on the tornado 5 eventloop issue. I need to read up on how atexit and asyncio interact.

**Comment by mattvw** [comment]
Any update on this? I'm running into this issue as well.

**Comment by sabarnwa** [comment]
Any updates please.

**Comment by minrk** [comment]
This is fixed by #2040
